src/peft/tuners/loha/layer.py

```diff
@@ -98,38 +106,158 @@ def reset_adapter_parameters_random(self, adapter_name: str):
+        
+        For LoHa with separate ranks: (w1a @ w1b) ⊙ (w2a @ w2b)
+        where w1a,w1b have rank r1 and w2a,w2b have rank r2.
+        We want this to approximate the pretrained weight W.
+        
+        Strategy: Use SVD to split the weight matrix, allocating r1 singular values
+        to the first component and r2 to the second component.
+        """
+        if adapter_name in self.hada_w1_a.keys():
+            base_layer = self.get_base_layer()
+            weight = base_layer.weight.data
+            
+            # Flatten weight for linear layers
+            if isinstance(base_layer, nn.Linear):
+                W = weight  # (out_features, in_features)
+            else:
+                # For conv layers, flatten to 2D
+                W = weight.reshape(weight.shape[0], -1)
+            
+            # Get the separate ranks
+            r1 = self.r1[adapter_name]
+            r2 = self.r2[adapter_name]
+            
+            try:
+                U, S, Vh = torch.linalg.svd(W, full_matrices=False)
+                
+                # Split singular values between r1 and r2
+                # Take top r1+r2 singular values and split them
+                total_r = min(r1 + r2, len(S))
+                actual_r1 = min(r1, total_r)
+                actual_r2 = min(r2, total_r)
+                
+                # Get components for first Hadamard term (rank r1)
+                U_r1 = U[:, :actual_r1]  # (m, r1)
+                S_r1 = S[:actual_r1]  # (r1,)
+                Vh_r1 = Vh[:actual_r1, :]  # (r1, n)
+                
+                # Get components for second Hadamard term (rank r2)
+                # Use next r2 singular values or reuse if not enough
+                if actual_r1 + actual_r2 <= len(S):
+                    U_r2 = U[:, actual_r1:actual_r1 + actual_r2]  # (m, r2)
+                    S_r2 = S[actual_r1:actual_r1 + actual_r2]  # (r2,)
+                    Vh_r2 = Vh[actual_r1:actual_r1 + actual_r2, :]  # (r2, n)
+                else:
+                    # Reuse early singular values if needed
+                    U_r2 = U[:, :actual_r2]  # (m, r2)
+                    S_r2 = S[:actual_r2]  # (r2,)
+                    Vh_r2 = Vh[:actual_r2, :]  # (r2, n)
+                
+                # Initialize first component: w1a @ w1b
+                # Use fourth root so that (w1a @ w1b) ⊙ (w2a @ w2b) ≈ W
+                fourth_root_S1 = torch.pow(S_r1, 0.25)
+                self.hada_w1_a[adapter_name].data.copy_(U_r1 * fourth_root_S1)
+                self.hada_w1_b[adapter_name].data.copy_(fourth_root_S1.unsqueeze(1) * Vh_r1)
+                
+                # Initialize second component: w2a @ w2b
+                fourth_root_S2 = torch.pow(S_r2, 0.25)
+                self.hada_w2_a[adapter_name].data.copy_(U_r2 * fourth_root_S2)
+                self.hada_w2_b[adapter_name].data.copy_(fourth_root_S2.unsqueeze(1) * Vh_r2)
+                
+            except Exception as e:
```

Comment:
IMO, there is no need to catch the exception. As a user, if I explicitly indicated that I want ABBA and it fails, I would rather get an error than a warning.